### PR TITLE
Log thread name and exception type on unhandled exceptions

### PR DESCRIPTION
**Pull request by seelabs**

### CODE REVIEW

**Comment by seelabs** [comment]
This PR was a joint effort of @HowardHinnant @scottschurr and myself. I added @miguelportilla to make sure we got a review from someone who wasn't involved in coding this.

**Comment by seelabs** [comment]
@scottschurr I pushed a commit that removes the lambda. I suppose we could save the lambda in a `thread_local std::function`, but I don't think that's worth it (I'm not sure when `std::function` allocates, and if it does it's probably about as expensive as just creating the string non-lazily).

**Comment by miguelportilla** [comment]
I like how this got simplified. :+1:

File: src/ripple/core/ThreadEntry.h [hunk 1]
```diff
@@ -30,6 +30,16 @@
...
+namespace detail
+{
+void setThreadName(std::string name);
+}
+
+void terminateHandler();
+#endif
+
 /**
 Report uncaught exceptions to DebugLog and cerr
```
Comment (scottschurr) [inline src/ripple/core/ThreadEntry.h]: Note that this comment block needs tending to.  In particular the paragraph starting on line 59 (regarding the lambda) should be removed.

File: src/ripple/core/ThreadEntry.h [hunk 2]
```diff
@@ -17,8 +17,8 @@
 */
 //==============================================================================
 
-#ifndef RIPPLE_CORE_REPORT_UNCAUGHT_EXCEPTION_H_INCLUDED
-#define RIPPLE_CORE_REPORT_UNCAUGHT_EXCEPTION_H_INCLUDED
+#ifndef RIPPLE_CORE_THREAD_ENTRY_H_INCLUDED
+#define RIPPLE_CORE_THREAD_ENTRY_H_INCLUDED
 
 #include <ripple/basics/Log.h>
 #include <boost/coroutine/exceptions.hpp> // forced_unwind exception
```
Comment (scottschurr) [inline src/ripple/core/ThreadEntry.h]: I'm guessing a number of the includes can be removed from this file.

Comment (HowardHinnant) [inline src/ripple/core/ThreadEntry.h]: But when you do remove the includes, add some of them back to ThreadEntry.cpp.  I'm thinking `<iostream>`, `<exception>`, `<ripple/basics/Log.h>` and `<string>` need to be seen by ThreadEntry.cpp.

**Comment by HowardHinnant** [comment]
👍  Looks good to me.  Tests out on macOS where all of this is disabled.  I'm in favor of Scott S' minor comments.  Thanks!

File: src/ripple/core/impl/ThreadEntry.cpp [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+}
+}
+
+void terminateHandler()
+{
+    if (std::current_exception())
+    {
+        try
+        {
+            throw;
```
Comment (scottschurr) [inline src/ripple/core/impl/ThreadEntry.cpp]: Now that I'm thinking harder about this, it's really interesting that we can do a `throw/catch` here in the `terminateHandler` without losing the stack trace.  There's clearly some magic going on.  It's probably related to the need to call the `terminateHandler` no matter the reason for exiting (e.g., even for an uncaught exception).

Comment (HowardHinnant) [inline src/ripple/core/impl/ThreadEntry.cpp]: That's a good observation Scott.  Details for the interested:

Exception throwing (on gcc and clang) is a 2-phase process.  Phase 1 walks the stack _without_ unwinding it (or executing destructors) looking for a matching catch clause.  When a matching catch clause is found, then phase 2 kicks in and actually unwinds the stack up to the matching catch, executing destructors along the way.  It is the second phase which erases the stack trace (up to the matching catch).  By placing a matching catch at the top of the thread stack, we ensure that we always erase the stack trace up to that top matching catch.

The good part is where phase 1 _doesn't_ find a matching catch.  In that case the compiler calls _down_ the stack to `std::terminate()`.  As `terminate()` is called, the entire stack above the throw is still intact.  `terminal()` then calls the handler (down the stack again).  In the handler we throw and catch all in the same tiny local area of the stack (still under the point of throw).  So the stack only gets unwound from this local `throw;` to this local `catch()`, which is to say, not at all.  After the handler completes, `std::abort()` is called.  And everything above the point of throw (and a little bit of stuff below the throw) is still on the stack, waiting for the abort process to log it.

Comment (scottschurr) [inline src/ripple/core/impl/ThreadEntry.cpp]: Thanks a bunch for the explanation, @HowardHinnant.  Makes perfect sense.

**Comment by scottschurr** [comment]
:+1:, but we should wait for the CI builds to finish.

**Comment by scottschurr** [comment]
CI is looking questionable.  Travis had an internal compiler error on the Coverage build.  Problems with thread local storage?

**Comment by vinniefalco** [comment]
You can use the `boost::thread_local` or whatever its called

**Comment by seelabs** [comment]
@vinniefalco The code that uses `thread_local` is disabled on OSX since it already has good unhandled exception reporting. Even if we used `boost::thread_local` we'd still keep it disabled on OSX, so we may as well use standard `thread_local`. (I assume @HowardHinnant knows better).

**Comment by HowardHinnant** [comment]
Yes, that's pretty much my thought process.  As long as `thread_local` works on linux and VS, I think we're good.

**Comment by scottschurr** [comment]
A rebuild of the earlier failed build is proceeding much better.  So apparently my concern about thread local storage was unfounded (or at least not deterministic 😉 .

**Pull request closed without merging**